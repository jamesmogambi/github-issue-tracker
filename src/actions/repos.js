require("isomorphic-fetch");

// SET_REPOSITORIES
export const setRepositories = (payload) => ({
  type: "SET_REPOSITORIES",
  payload,
});
// CLEAR_REPOSITORIES
export const clearRepositories = () => ({
  type: "CLEAR_REPOSITORIES",
});

// ERROR
export const error = (payload) => ({
  type: "ERROR",
  payload,
});

// Get Github Repositories
export const getRepositories = (params) => {
  return (dispatch) => {
    dispatch(clearRepositories());
    fetch("https://api.github.com/graphql", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        authorization: `bearer ${process.env.GITHUB_ACCESS_TOKEN}`,
      },
      body: JSON.stringify({
        query: `
               query($github_login: String!) {
    user(login: $github_login) {
      name
      repositories(first: 100) { 
        nodes {
             id
        name
        description 
        url 
        updatedAt
        forkCount
        }
      }
    }
  }
                 `,
        variables: {
          github_login: params.username,
        },
      }),
    })
      .then((res) => res.json())
      .then((res) => {
        if (res.data.user) {
          let repos = res.data.user.repositories.nodes;
          let owner = res.data.user.name;
          dispatch(
            setRepositories({
              repositories: repos,
              repositoryOwner: params.username,
            })
          );
        }
      })
      .catch((err) => {
        dispatch(error(err));
        console.error(err);
      });
  };
};
