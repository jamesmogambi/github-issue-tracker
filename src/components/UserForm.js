import React from "react";

export default class UserForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      username: "",
    };
  }
  onUserNameChange = (e) => {
    const username = e.target.value;
    this.setState(() => ({ username }));
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (this.state.username) {
      this.props.onSubmit({
        username: this.state.username,
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        <input
          type="text"
          placeholder="Github username"
          autoFocus
          className="text-input"
          value={this.state.username}
          onChange={this.onUserNameChange}
        />

        <div>
          <button className="button">View Repos</button>
        </div>
      </form>
    );
  }
}
