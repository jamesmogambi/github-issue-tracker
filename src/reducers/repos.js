// Repositories Reducer
const initialState = {
  repositories: [],
  repositoryOwner: null,
  error: null,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case "SET_REPOSITORIES":
      const { repositories, repositoryOwner } = action.payload;
      return {
        ...state,
        repositoryOwner,
        repositories,
      };
    case "CLEAR_REPOSITORIES":
      return {
        repositories: [],
        repositoryOwner: null,
        error: null,
      };

    default:
      return state;
  }
};
