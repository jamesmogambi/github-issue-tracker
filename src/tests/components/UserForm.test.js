import React from "react";
import { shallow } from "enzyme";
import moment from "moment";
import UserForm from "../../components/UserForm";
import repos from "../fixtures/repos";

test("should render UserForm correctly", () => {
  const wrapper = shallow(<UserForm />);
  expect(wrapper).toMatchSnapshot();
});

test("should set username on input change", () => {
  const value = "James Edwards";
  const wrapper = shallow(<UserForm />);
  wrapper.find("input").at(0).simulate("change", {
    target: { value },
  });
  expect(wrapper.state("username")).toBe(value);
});

test("should render for invalid form submission", () => {
  const wrapper = shallow(<UserForm />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {},
  });
  expect(wrapper.state("username").length).toEqual(0);
  expect(wrapper).toMatchSnapshot();
});

test("should call onSubmit prop for valid form submission", () => {
  const data = "James Edwards";
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(<UserForm user={data} onSubmit={onSubmitSpy} />);
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {},
  });
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    username: data,
  });
});
