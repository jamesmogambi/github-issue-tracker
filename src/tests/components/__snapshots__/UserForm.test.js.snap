// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render UserForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Github username"
    type="text"
    value=""
  />
  <div>
    <button
      className="button"
    >
      View Repos
    </button>
  </div>
</form>
`;

exports[`should render for invalid form submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Github username"
    type="text"
    value=""
  />
  <div>
    <button
      className="button"
    >
      View Repos
    </button>
  </div>
</form>
`;

exports[`should render for invalid form submission 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Github username"
    type="text"
    value=""
  />
  <div>
    <button
      className="button"
    >
      View Repos
    </button>
  </div>
</form>
`;
